docs/source-fabric/advanced/compile.rst

```diff
@@ -0,0 +1,299 @@
+        t0 = time.time()
+        compiled_model(input)
+        torch.cuda.synchronize()
+        print(f"2nd forward: {time.time() - t0:.2f} seconds.")
+
+    With ``automatic_dynamic_shapes=True``:
+
+    .. code-block:: text
+
+        1st forward: 41.90 seconds.
+        2nd forward: 89.27 seconds.
+
+    With ``automatic_dynamic_shapes=False``:
+
+    .. code-block:: text
+
+        1st forward: 42.12 seconds.
+        2nd forward: 47.77 seconds.
+
+    Numbers produced with NVIDIA A100 SXM4 40GB, PyTorch 2.2.0, CUDA 12.1.
+
+----
+
+
+***********************************
+Experiment with compilation options
+***********************************
+
+There are optional settings that, depending on your model, can give additional speedups.
+
+**CUDA Graphs:** By enabling CUDA Graphs, CUDA will record all computations in a graph and replay it every time forward and backward is called.
+The requirement is that your model must be static, i.e., the input shape must not change and your model must execute the same operations every time.
+Enabling CUDA Graphs often results in a significant speedup, but sometimes also increases the memory usage of your model.
+
+.. code-block:: python
+
+    # Enable CUDA Graphs
+    compiled_model = torch.compile(model, mode="reduce-overhead")
+
+    # This does the same
+    compiled_model = torch.compile(model, options={"triton.cudagraphs": True})
+
+|
+
+**Shape padding:** The specific shape/size of the tensors involved in the computation of your model (input, activations, weights, gradients, etc.) can have an impact on the performance.
+With shape padding enabled, ``torch.compile`` can extend the tensors by padding to a size that gives a better memory alignment.
+Naturally, the tradoff here is that it will consume a bit more memory.
+
+.. code-block:: python
+
+    # Default is False
+    compiled_model = torch.compile(model, options={"shape_padding": True})
+
+
+You can find a full list of compile options in the `PyTorch documentation <https://pytorch.org/docs/stable/generated/torch.compile.html>`_.
+
+----
+
+
+*******************************************************
+(Experimental) Apply torch.compile over FSDP, DDP, etc.
+*******************************************************
+
+As stated earlier, we recommend that you compile the model before calling ``fabric.setup()``.
+However, if you are using DDP or FSDP with Fabric, the compilation won't incorporate the distributed calls inside these wrappers by default.
+In an experimental feature, you can let ``fabric.setup()`` reapply the ``torch.compile`` call after the model gets wrapped in DDP/FSDP internally.
+In the future, this option will become the default.
+
+.. code-block:: python
+
+    # Choose a distributed strategy like DDP or FSDP
+    fabric = L.Fabric(devices=2, strategy="ddp")
+
+    # Compile the model
+    model = torch.compile(model)
+
+    # Default: `fabric.setup()` will not reapply the compilation over DDP/FSDP
```

Comment:
Since we are usually quite sensitive with breaking changes, I wanted to make this experimental first for one release. I still need to do the work and run a batch of lit-gpt models with this settings to be sure this is not counter productive as a default.

Reply:
I don't expect any significant number of users to find and enable this private flag. If you are afraid because the release is close, that's fine. But I would enable it as soon as we can after so that at least nightly users try it out